Lib/unittest/case.py

```diff
@@ -170,6 +171,46 @@ def _is_subtype(expected, basetype):
         return all(_is_subtype(e, basetype) for e in expected)
     return isinstance(expected, type) and issubclass(expected, basetype)
 
+
+def _heuristic_diff(a: list[str], b: list[str]) -> Iterator[str]:
+    """After testing the magnitude of the inputs, preferably return the output
+    of difflib.ndiff, but fallback to difflib.unified_diff for prohibitively
+    expensive inputs. How cost is calculated:
+
+    Cost is calculated according to this heuristic:
+
+        cost = (number of differing lines
+                * total length of all differing lines)
+
+    This heuristic is used because the time complexity of ndiff is
+    approximately O((diff)^2), where `diff` is the product of the number of
+    differing lines, and the total length of differing lines. On the other
+    hand, unified_diff's cost is the same as the cost of producing `diff`
+    by itself: O(a + b).
```

Comment:
Note that this isn't true. Both have worst-case quadratic time, but on different kinds of input, and it's less _likely_ to get provoked in the simpler kind of differencing `unified_diff` tries.